### ISSUE TITLE
Possible dependency issues (upgrading is not easy); mn2pdf fatal errors including HTML

### ISSUE THREAD

**Issue by jerstlouis** [metanorma/metanorma#437]
I was having some issues compiling latest [T20-GDC D020 ER](https://gitlab.ogc.org/ogc/T20-GDC/-/tree/main/D020) with metanorma:

`org.apache.fop.fo.ValidationException: "fo:table-row" is not a valid child of "fo:table"! (See position 428:100)`

so I decided to upgrade (`gem install metanorma metanorma-ogc metanorma-cli`), which resulted in new errors, which seemed to indicate a missing dependency:

`~/.local/share/gem/ruby/3.2.0/gems/lutaml-model-0.3.9/lib/lutaml/model/xml_mapping.rb:39:in `map_element': unknown keyword: :render_default (ArgumentError)`.

After `gem install gem install lutaml-model`, I now have a new errors:

```
~/.local/share/gem/ruby/3.2.0/gems/metanorma-standoc-2.10.3/lib/metanorma/standoc/converter.rb:28:in `block in <class:Converter>': uninitialized constant Metanorma::Plugin::Lutaml::LutamlEaXmiPreprocessor (NameError)

        preprocessor Metanorma::Plugin::Lutaml::LutamlEaXmiPreprocessor
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^
Did you mean?  Metanorma::Plugin::Lutaml::LutamlPreprocessor
```

for which I can't seem to figure out what missing dependency I should manually be updating.
I did try `gem install metanorma-plugin-lutaml` and `lutaml-xmi` with no luck.

EDIT: It seems I had to `gem install metanorma-standoc`. Why are there so many metanorma gems without dependencies between them? Which one should I be manually installing when upgrading?

Now I am back to the original compiling issue with T20-GDC D020:

```
[mn2pdf] [FATAL] "fo:table-cell" is not a valid child of "fo:block"! (See position 128:168)
[mn2pdf] Can't obtain table's widths information: javax.xml.transform.TransformerException: org.apache.fop.fo.ValidationException: "fo:table-cell" is not a valid child of "fo:block"! (See position 128:168)
```

Thanks for any assistance.

EDIT 2: Things stopped compiling since b64fc6372623939e9127e5cdd0928de667058eb1 (in our T20 repo) when an HTML file has been included directly in the document. This is probably not a good idea and mostly on us, but a more user-friendly explanation of the problem would have been helpful.

Low priority since this is no longer blocking.

**Comment by opoudjis** [comment]
> EDIT: It seems I had to gem install metanorma-standoc. Why are there so many metanorma gems without dependencies between them? Which one should I be manually installing when upgrading?

The metanorma-standoc gemspec has

````
  spec.add_dependency "metanorma-plugin-datastruct", "~> 0.3.0"
  spec.add_dependency "metanorma-plugin-glossarist", "~> 0.2.0"
  spec.add_dependency "metanorma-plugin-lutaml", "~> 0.7.0"
````

And Metanorma::Plugin::Lutaml::LutamlEaXmiPreprocessor is an artefact of metanorma-plugin-lutaml . The dependency is there in black and white, and that dependency and the class involved has not been touched for months.

I cannot tell what happened in your installation, but Metanorma has HUNDREDS of gems in its stack, and it is courting disaster to update one gem in isolation. What you should be updating is 

````
bundle update
````

with a Gemfile consisting of

````
source "https://rubygems.org"

gem "metanorma-cli"
````

which inherits the entire stack. That is what we do with all mn-samples-* repositories, such as mn-samples-ogc

That error is a PDF processing error: it clearly did not expect to find table-row and table-cell where it did. (Those aren't Metanorma XML elements, so I would have to defer to you on how they ended up in the XML it was processing.)

It's hard to see how we could forestall this: we can abort if the XML contains elements it won't recognise—but there are chunks of the XML where such extraneous XML is expected, and making the abort friendly is going to be very difficult, when its provenance is going to be unclear to begin with. (You already have the message after all from mn2pdf on what tag it did not expect to see there.)

**Comment by jerstlouis** [comment]
@opoudjis Thanks for the help.

> bundle update

So I figured I had to do the following (I am not Ruby-savvy at all):
- Create a directory containing a file called `Gemfile` with that content
- Call `bundle config set --local path ~/.local/share/gem/ruby/3.2.0/gems/` so that gems get installed at the usual place that `gem install` would install there before (otherwise it tried to install them system-wide)
- Then `bundle update` seemed to work. I'll try using this from now on.

>  Metanorma has HUNDREDS of gems in its stack, and it is courting disaster to update one gem in isolation

My understanding was that `gem install` upgrades all dependencies as needed (isn't that the case? it did seem to update several dependencies automatically), so I was really not thinking of it as updating _one gem in isolation_. I thought that if I do `gem install metanorma-cli`, assuming all dependencies are set up properly, things should just work -- otherwise this is a dependency bug that should be reported and fixed.

> Those aren't Metanorma XML elements, so I would have to defer to you on how they ended up in the XML it was processing

The document was trying to use an [HTML passthrough block](https://docs.asciidoctor.org/asciidoc/latest/pass/pass-block/), which is something that is supported by Asciidoctor:
```
++++
include::02-topics-gdc-table.html[]
++++
```
where the content of that HTML file included `<table>` and `<style>` tags (directly, without any outer `<html>` element).

Is that also supported in Metanorma? The HTML output did get generated successfully, but the HTML appeared verbatim rather than as a table.
It probably would not have worked as expected in PDF, but it really should not result in broken XML causing a fatal error in _mn2pdf_.

The use case was to set up background color for table cells.

Thanks again.

Wishing you and the whole Metanorma team all the best for 2025! :)

**Comment by opoudjis** [comment]
I've been away for several weeks.

````
The document was trying to use an [HTML passthrough block](https://docs.asciidoctor.org/asciidoc/latest/pass/pass-block/), which is something that is supported by Asciidoctor:

++++
include::02-topics-gdc-table.html[]
++++
where the content of that HTML file included <table> and <style> tags (directly, without any outer <html> element).

Is that also supported in Metanorma? The HTML output did get generated successfully, but the HTML appeared verbatim rather than as a table.
````

No. And it is not, because (a) whereas Asciidoctor can assume that it is only generating HTML, Metanorma generates all of Metanorma Semantic XML, Metanorma Presentation XML, HTML, DOC, and PDF, (b) passthrough is interpreted as literal code, not XML code, unless a format is specified, and (c) the include is going to sidestep any passthrough processing to a specific format anyway.

This will not work: it is antithetical to Metanorma's processing model, which generates Metanorma XML, not HTML, in its first pass. What *will* work, I believe, is:

````
include::02-topics-gdc-table.html[]
````

and in 02-topics-gdc-table.html:

````
[format=html]
++++
(table markup)
++++
````

without the `[format=html]` wrapping the HTML markup, it will be treated as raw XML, and escaped at that.

Could you share the document you're working on? I have a bad feeling that there will still be things that go wrong.

**Comment by jerstlouis** [comment]
@opoudjis Sorry was very busy and forgot to get back to you on this issue.

In the document ( [Testbed 20 GDC](https://gitlab.ogc.org/ogc/T20-GDC/-/tree/main/D020) ) we moved away from this HTML table to an ASCIIDoc table (had to give up the colors, but I see there's already) #446 about this.

Still, I suggest the following could be addressed:
- Report an error early on if users try to include this `.html[]` outside of this `[format=html]` so that incomprehensible errors don't show up later, perhaps with a suggestion making users aware of this `[format=html]` possibility if it is a proper solution
- Fixing the dependency of `metanorma` on `metanorma-standoc` (if there is one?), and generally keeping dependencies correct so that people **can** install / upgrade metanorma with simply `gem install metanorma-cli` in addition to the `bundle update` method.

Other than this I believe this issue can be closed. Thank you!

**Comment by opoudjis** [comment]
You're asking in dot point 1 for us to work out whether some arbitrary XML is Metanorma XML or not.  We could try parsing it against the Metanorma XML grammar, except that is going to raise way too many errors.

After a couple of hours' work, I am managing to parse a fragment of XML against the Metanorma XML schema. I will only reject the XML if it contains extraneous elements, and that will trigger an abort.

I will not be suggesting that it might be HTML instead, because we don't just allow HTML, and it would be a complete waste of effort to infer what the correct format is; but I will advise that users should be using the format attribute.

This has been a very large amount of effort for an edge case.

**Comment by jerstlouis** [comment]
>  You're asking in dot point 1 for us to work out whether some arbitrary XML is Metanorma XML or not.

Sorry, I'm not familiar at all with the inner workings of metanorma, but my suggestison was simply to detect early on (potentially before any XML is generated at all -- in fact I wish metanorma had nothing at all to do with XML).

Essentially my suggestion was to look out for includes in the .adoc that use a .html extension -- something that works with ASCIIDoc, but not metanorma.

> This has been a very large amount of effort for an edge case.

Sorry to trigger this large amount of efforts, but I think anything that avoids intense frustration and wasted hours for users is well worth the effort :)

Thanks for working on this.

**Comment by opoudjis** [comment]
The processing has to be more general than that; for example we have to deal with inline examples like

````
[pass]
++++
<table>
...
</table>
++++
````

and waiting on an .html suffix in an included file won't deal with that. Nor for that matter would we catch cases where you've specified an .xml suffix, but the content in the .xml file is illegal. That's not going to be enough, nor robust enough.

I'm not overjoyed I've ended up spending 5 hours on this, but it is a legitimate thing to try to prevent. The implementation I've just finished is:

* extract the XML in a pass block (including external files), but only if no format is specified (which means we expect it to be Metanorma XML)
* parse that XML fragment against a grammar derived from the Metanorma grammar, but requiring the XML fragment root element to be the start element of the grammar. (That's how we can parse for just the fragment.)
* if we find extraneous elements in it, we conclude that this is not Metanorma XML, and we abort: the resulting XML will never be valid, and there is no point continuing
* however, we won't abort if the XML is incomplete (not well formed), which is possible if this is an XML fragment (e.g. we specify the start and end of a block in two separate pass spans.)

I have testing in place for this update, and it will go out with the next release of Metanorma.

**Comment by opoudjis** [comment]
In order to validate document fragments against Relax NG schemas, we need to remove comments from the root of all such schemas. This is an old libxml2 bug, which has not been fixed in close to two decades.

**Comment by opoudjis** [comment]
Also, need to parameterise schema location, so that downstream we validate fragments against current flavour's schema.

**Comment by opoudjis** [comment]
Had not been testing running of local schema, which is needed to catch schema errors.

**Comment by jerstlouis** [comment]
>  The dependencies of metanorma are explicitly set, and the failure of metanorma-ogc to update is an issue external to the gem config, which I'm not sure we can address, short of wiping and regenerating Gemfile.lock

I believe there really was a missing dependency on metanorma-standoc (from metanorma or metanorma-cli), but perhaps that was fixed since. How can I verify those dependencies again?

Thanks for all the hard work!

**Comment by opoudjis** [comment]
The more dependencies are specified, the more disasters can happen with dependencies, and the proper dependency is of metanorma-cli on metanorma and on metanorma-standoc independently.

**Comment by opoudjis** [comment]
Need to specify that this is ONLY done for Metanorma XML passthrough, not other formats.
